### PR TITLE
[MediaRecorder] timecode provided in BlobEvent is incorrect

### PR DESCRIPTION
**Pull request by jyavenard**
#### ced7f5e2066e77bb308efb5274a69119421cfd81
<pre>
[MediaRecorder] timecode provided in BlobEvent is incorrect
<a href="https://bugs.webkit.org/show_bug.cgi?id=282320">https://bugs.webkit.org/show_bug.cgi?id=282320</a>
<a href="https://rdar.apple.com/138895100">rdar://138895100</a>

Reviewed by Anne van Kesteren.

The timecode returned was always of the time at which requestData() was called and not
&quot;For a MediaRecorder instance, the timecode in the first produced BlobEvent MUST contain 0.
Subsequent BlobEvent&apos;s timecode contain the difference of the timestamp of creation of the
first chunk in said BlobEvent and the timestamp of the first chunk of the first produced BlobEvent&quot;

imported wpt&apos;s MediaRecorder-blob-timecode.https.html to test this functionality.
The test exposed an issue in our implementation (which the test didn&apos;t deal with).
It sets a timeslice of 0 and assume that we would always get a monotonically increasing
timecode, but if we attempts to make 0ms long segment that may not always be the case.

As a workaround, and considering real-life usage, if a timeslice is set and no data was recorded at the time
the timer occurs, we will not emit `dataavailable` event. We also limit how small the timeslice can be to 100ms.

On platforms using AVFoundation, we make the minimum segment size to be 1s. Using a MP4 recorder calling requestData()
will return empty Blobs if called before 1s of data has been accumulated due to the structure of ISOBMFF (it needs to
write a complete samples table first). An AVFormatWriter will also errors with &quot;NotReady&quot; if we attempt to generate
too short segments.
With WebM data is written as received, but a new cluster will always be at least one second long.

We also hold off writing any samples until the AudioSampleBufferCompressor has started to return audible data to
ensure video and audio frames are already interleaved in monotonically increasing time order.

* LayoutTests/imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-blob-timecode.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-blob-timecode.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-record/w3c-import.log:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Modules/mediarecorder/MediaRecorder.cpp:
(WebCore::MediaRecorder::MediaRecorder):
(WebCore::MediaRecorder::startRecording):
(WebCore::MediaRecorder::requestData):
(WebCore::MediaRecorder::requestDataInternal): Split method. Don&apos;t emit dataavailable event unless we actually have new data.
* Source/WebCore/Modules/mediarecorder/MediaRecorder.h: Set a minimum timeslice of 0.1s.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::takeData):
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedAudioSampleBuffers):
(WebCore::MediaRecorderPrivateEncoder::flushToEndSegment): Return pair returnCode/time instead of setting a member variable.
Test exposed a case where we could have not written all available frames.
(WebCore::MediaRecorderPrivateEncoder::muxNextFrame):
(WebCore::MediaRecorderPrivateEncoder::fetchData):
(WebCore::MediaRecorderPrivateEncoder::flushPendingData):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:

Canonical link: <a href="https://commits.webkit.org/286442@main">https://commits.webkit.org/286442@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/14773f709983a768189da0401435ec46367b8883

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75973 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55002 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28872 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80470 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27239 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78089 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64144 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3298 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59572 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17728 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79040 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49452 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65244 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39928 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46852 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22728 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25566 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67968 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23065 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81934 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3342 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2125 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67798 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3496 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65212 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67108 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16723 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11056 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9178 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3292 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6098 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3313 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4251 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3320 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp [hunk 1]
```diff
@@ -905,11 +905,12 @@ void MediaRecorderPrivateEncoder::fetchData(CompletionHandler<void(RefPtr<Fragme
         return flushPendingData(currentTime)->whenSettled(queueSingleton(), [protectedThis, this, completionHandler = WTFMove(completionHandler), currentTime]() mutable {
             assertIsCurrent(queueSingleton());
             RefPtr data = takeData();
```
Comment (annevk) [inline Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp]: Drive-by: since it's assumed data is non-null, would it be cleaner to assign to a Ref?

**Review by annevk — APPROVED** [review]

File: LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-timecode-expected.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+
+PASS BlobEvent timecode increases
```
Comment (youennf) [inline LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-timecode-expected.txt]: Can you import WPT mediacapture-record/MediaRecorder-blob-timecode.https.html instead?

**Review by annevk — APPROVED** [review]
Nice, still looks good (and better because more WPT!), but I suspect this is going to fail on the bots because the test lacks an -expected.txt file.

**Comment by jyavenard** [comment]
The wpt tests appear rather incorrect. It sets a timeslice of 0, so we are to simulate a call to requestData every 0ms. 

It asserts then that each timecode is monotonically increasing which assumes that we do have data being returned all the time; that is not always the case. It should handle the case where after 0ms we not have more data being ready.

The spec states:
```
If timeslice is not undefined, then once a minimum of timeslice milliseconds of data have been collected, or some minimum time slice imposed by the UA, whichever is greater, start gathering data into a new [Blob](https://w3c.github.io/FileAPI/#dfn-Blob) blob, and queue a task, using the DOM manipulation task source, that [fires a blob event](https://w3c.github.io/mediacapture-record/#fire-a-blob-event) named [dataavailable](https://w3c.github.io/mediacapture-record/#eventdef-mediarecorder-dataavailable) at recorder with blob.
```

we do not have a minimum time, but we could set one, like 1s segment minimum which would be more reasonable that making the system being peg calling continuously MediaRecorderPrivate::fetchData constantly.

**Comment by jyavenard** [comment]
chrome uses a minimum timeslice duration of 1s (even though their comments sets 10Hz) https://source.chromium.org/chromium/chromium/src/+/main:media/muxers/mp4_muxer.cc;l=16;drc=0afc9ac9afcaab79fc54299039f4d27abf3a086d

**Comment by philn** [comment]
> chrome uses a minimum timeslice duration of 1s (even though their comments sets 10Hz) https://source.chromium.org/chromium/chromium/src/+/main:media/muxers/mp4_muxer.cc;l=16;drc=0afc9ac9afcaab79fc54299039f4d27abf3a086d

Their webm muxer uses 100ms clusters.

**Comment by jyavenard** [comment]
> > chrome uses a minimum timeslice duration of 1s (even though their comments sets 10Hz) https://source.chromium.org/chromium/chromium/src/+/main:media/muxers/mp4_muxer.cc;l=16;drc=0afc9ac9afcaab79fc54299039f4d27abf3a086d
> 
> Their webm muxer uses 100ms clusters.

yes they do there.

Firefox don't set a minimum, but they will not write a segment smaller than the timeslice, so they don't use a timer like we do.

I think this test will only pass consistently with Chrome. I will lodge a bug on wpt.

Cocoa implementation was designed so that it will only ever create a segment that is complete and playable (not 100% per spec, but that was our decision), e.g. it can't generate an empty segment, this will cause the promise to not be resolved and the test to timeout. 

It did expose an issue in Apple's implementation however, as we flush super quickly we write all the audio we have, a pending video frame is then appended with a time that is earlier than audio and after we've completed the previous ISOBMFF segment: this makes the AVAssetWriter errors and it stops outputting data.

**Comment by jyavenard** [comment]
I created https://github.com/web-platform-tests/wpt/issues/49088

**Comment by jyavenard** [comment]
LayoutTests/fast/mediacapturefromelement/CanvasCaptureMediaStream-offscreencanvas.html now fails because it sets a timeslice of 1s ; but only ever paint once on the canvas. So we won't get a video duration greater than 1s for data to be output. Would need to call `stop` on the recorder first.

**Comment by jyavenard** [comment]
WPE crashes are ```
RealtimeMediaSource& MediaStreamTrackPrivate::source()
{
    ASSERT(isMainThread());
    return m_sourceObserver->source();
}
```

assertion as this isn't on the main thread.

**Comment by philn** [comment]
Looks like a regression in main, I'm taking a look.

**Comment by darinadler** [comment]
Looks like those WPE crashes are resolved now, EWS all green
